### PR TITLE
Telemetry

### PR DESCRIPTION
**Pull request by siggisim**

### CODE REVIEW

File: server/telemetry/telemetry_client.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+		printIfVerbose("Error generating telemetry log UUID: %s", err)
+		return unknownFieldValue
+	}
+	return logUUID.String()
+}
+
+func getHostname() string {
+	hostname, err := os.Hostname()
+	if err != nil {
+		printIfVerbose("Error retrieveing hostname: %s", err)
```
Comment (tylerwilliams) [inline server/telemetry/telemetry_client.go:212]: retrieving

File: server/telemetry/telemetry_client.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+		printIfVerbose("Error posting telemetry data: %s", err)
+		t.failedLogs = append(t.failedLogs, log)
+		return
+	}
+
+	t.failedLogs = []*telpb.TelemetryLog{}
+	printIfVerbose("Telemetry data posted: %+v", response)
+}
+
+func printIfVerbose(message string, args ...interface{}) {
```
Comment (tylerwilliams) [inline server/telemetry/telemetry_client.go:129]: TODO(tylerw): use glog or similar. Then we can just do log.Verbose(blah) here.

File: CONFIG.md [hunk 3]
```diff
@@ -263,10 +263,17 @@ integrations:
 #### APIKey
 ```api_key:``` *Optional* The default API key to use for on-prem enterprise deploys with a single organization/group.
 
+# Telemetry
+
+At BuildBuddy, we collect telemetry for the purpose of helping us build a better BuildBuddy. Data about how BuildBuddy is used is collected to better understand what parts of BuildBuddy needs improvement and what features to build next. Telemetry also helps our team better understand the reasons why people use BuildBuddy and with this knowledge we are able to make better product decisions.
+
+The telemetry data we collect is reported once per day and contains only aggregate stats like invocation counts and feature usage information. Our telemetry infrastructure is also used to report when important security updates are available. For a complete picture of the telemetry data we collect, you can see our [telemetry proto](https://github.com/buildbuddy-io/buildbuddy/blob/master/proto/telemetry.proto).
+
+While we encourage users to enable telemetry collection and strive to be fully transparent about the data we collect to improve BuildBuddy, telemetry can be disabled at any time using the flag `--disable_telemetry=true`.
```
Comment (tylerwilliams) [inline CONFIG.md:272]: nice

File: proto/telemetry.proto [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  string app_version = 5;
+
+  // The URL at which BuildBuddy is running.
+  string app_url = 6;
+
+  // The hostname of this BuildBuddy instance.
+  string hostname = 7;
+
+  // Daily BuildBuddy usage statistics.
+  TelemetryStat daily_stat = 8;
```
Comment (tylerwilliams) [inline proto/telemetry.proto:52]: sure you don't want to just call the fields telemetry_stat and telemetry_feature?

File: server/util/db/db.go [hunk 5]
```diff
@@ -51,3 +57,10 @@ func GetConfiguredDatabase(c *config.Configurator) (*DBHandle, error) {
 	}
 	return nil, fmt.Errorf("No database configured -- please specify at least one in the config")
 }
+
+func (d *DBHandle) StartOfDayTimestamp(offset int) string {
```
Comment (tylerwilliams) [inline server/util/db/db.go:61]: Could we call this offsetDays or offsetPastDays or something cleverer / clearer than i'm suggesting?

File: server/telemetry/telemetry_client.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+	}
+
+	response, err := client.LogTelemetry(ctx, req)
+	if err != nil || response.Status.Code != 0 {
+		printIfVerbose("Error posting telemetry data: %s", err)
+		t.failedLogs = append(t.failedLogs, log)
+		return
+	}
+
+	t.failedLogs = []*telpb.TelemetryLog{}
```
Comment (tylerwilliams) [inline server/telemetry/telemetry_client.go:125]: We might want to add a check somewhere that this isn't growing without bound. Wdyt? You could just treat it as a fixed size queue and if it's more than 100 or something just start dropping old events.

File: server/util/db/db.go [hunk 7]
```diff
@@ -51,3 +57,10 @@ func GetConfiguredDatabase(c *config.Configurator) (*DBHandle, error) {
...
+	if d.dialect == sqliteDialect {
```
Comment (tylerwilliams) [inline server/util/db/db.go:62]: Can we do this all in go and remove the dialect-switching?

File: server/telemetry/telemetry_client.go (hunk 1 above)
Comment (siggisim) [reply server/telemetry/telemetry_client.go:212]: Done

File: proto/telemetry.proto (hunk 4 above)
Comment (siggisim) [reply proto/telemetry.proto:52]: Done

File: CONFIG.md (hunk 3 above)
Comment (siggisim) [reply CONFIG.md:272]: ✨

File: server/telemetry/telemetry_client.go (hunk 2 above)
Comment (siggisim) [reply server/telemetry/telemetry_client.go:129]: Done

File: server/util/db/db.go (hunk 5 above)
Comment (siggisim) [reply server/util/db/db.go:61]: Not needed anymore

File: server/util/db/db.go (hunk 7 above)
Comment (siggisim) [reply server/util/db/db.go:62]: Done

File: server/telemetry/telemetry_client.go (hunk 6 above)
Comment (siggisim) [reply server/telemetry/telemetry_client.go:125]: Done

**Pull request merged**